### PR TITLE
Implement the Synapse Handler for instrumentation of metrics

### PR DESCRIPTION
**Pull request by Nipunich**
## Purpose
Implement the Synapse Handler for instrumenting Prometheus metrics.

### CODE REVIEW

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/publish/SynapseObservabilityHandler.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+                incrementCount(inboundEndpointName);
+                startTimers(synCtx, inboundEndpointName, PrometheusMetricsConstants.INBOUND_ENDPOINT,
+                        null);
+            } else {
+                String context = axis2MessageContext.getProperty(PrometheusMetricsConstants.TRANSPORT_IN_URL).
+                        toString();
+                String apiInvocationUrl = axis2MessageContext.getProperty(PrometheusMetricsConstants.SERVICE_PREFIX).
+                        toString() + context.replaceFirst(DELIMITER, EMPTY);
+                String apiName = getApiName(context, synCtx);
```
Comment (Nirothipan) [inline components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/publish/SynapseObservabilityHandler.java:99]: Could you check in current SP analytics , how we are getting this api name .. This processing seems too big

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/util/PrometheusMetricsConstants.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+package org.wso2.micro.integrator.prometheus.observability.metrics.util;
+
+/**
+ * Constants required in instrumenting Prometheus metrics.
+ */
+public class PrometheusMetricsConstants {
+
+   // public static final String PROXY_NAME = "proxy.name";
+    public static final String SYNAPSE_REST_API = "SYNAPSE_REST_API";
```
Comment (sajinieKavindya) [inline components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/util/PrometheusMetricsConstants.java:27]: Shall we remove unnecessary comments?

File: pom.xml [hunk 3]
```diff
@@ -1213,6 +1213,38 @@
...
+                <groupId>io.prometheus</groupId>
+                <artifactId>simpleclient_hotspot</artifactId>
+                <version>${prometheus.version}</version>
+            </dependency>
+            <dependency>
+            <groupId>org.wso2.ei</groupId>
+            <artifactId>org.wso2.micro.integrator.core</artifactId>
+            <version>${project.version}</version>
+            <scope>compile</scope>
+        </dependency>
```
Comment (sajinieKavindya) [inline pom.xml:1247]: some indentation issues here.

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/pom.xml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Axis2Module>${project.artifactId}-${project.version}</Axis2Module>
+                        <Export-Package>
+                            org.wso2.micro.integrator.prometheus.observability.* ; version="${project.version}"
```
Comment (sajinieKavindya) [inline components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/pom.xml:76]: Shall we remove these unnecessary spacings in between package name and version?

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/pom.xml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                        </Export-Package>
+                        <Import-Package>
+                            org.wso2.carbon.inbound.endpoint.internal.http.api ; version="${project.version}",
```
Comment (sajinieKavindya) [inline components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/pom.xml:79]: from here too.

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/util/PrometheusMetricsConstants.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    public static final String PROXY_LATENCY_TIMER = "PROXY_LATENCY_TIMER";
+    public static final String API_LATENCY_TIMER = "API_LATENCY_TIMER";
+    public static final String INBOUND_ENDPOINT_LATENCY_TIMER = "INBOUND_ENDPOINT_LATENCY_TIMER";
+
+    public static final String HTTP_PORT = "http.nio.port";
+    public static final String JAVA_VERSION = "java.vm.specification.version";
+    public static final String JAVA_HOME = "java.home";
+    public static String PORT_OFFSET = "portOffset";
+    public static String INTERNAL_HTTP_API_PORT = "internal.http.api.port";
```
Comment (sajinieKavindya) [inline components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/util/PrometheusMetricsConstants.java:40]: can we make these two variables also final?

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/util/PrometheusMetricsConstants.java (hunk 2 above)
Comment (Nipunich) [reply components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/util/PrometheusMetricsConstants.java:26]: Fixed

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/pom.xml (hunk 4 above)
Comment (Nipunich) [reply components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/pom.xml]: Fixed

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/pom.xml (hunk 5 above)
Comment (Nipunich) [reply components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/pom.xml]: Fixed

File: components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/util/PrometheusMetricsConstants.java (hunk 6 above)
Comment (Nipunich) [reply components/mediation/data-publishers/org.wso2.micro.integrator.prometheus.observability/src/main/java/org/wso2/micro/integrator/prometheus/observability/metrics/util/PrometheusMetricsConstants.java]: Fixed

File: pom.xml (hunk 3 above)
Comment (Nipunich) [reply pom.xml]: Fixed

**Comment by Nipunich** [comment]
The modified implementation can be found in https://github.com/wso2/micro-integrator/pull/1662.

**Comment by Nipunich** [comment]
This Handler impelmenation was changed to support a Common Metric Implementation without binding to Prometheus.
This is addressed in https://github.com/wso2/micro-integrator/pull/1662

**Pull request closed without merging**